torchdynamo/symbolic_convert.py

```diff
@@ -490,31 +508,40 @@ def IMPORT_NAME(self, inst):
         fromlist = fromlist.as_python_constant()
         module_name = inst.argval
 
-        value = __import__(
-            module_name,
-            fromlist=fromlist,
-            level=level,
-            globals=self.f_globals,
-        )
-
-        if level != 0:
-            pkg = self.calc_package()
-            module_name = self.resolve_name(module_name, pkg, level)
-
-        # For __import__, when the name variable is of the form package.module,
-        # normally, the top-level package (the name up till the first dot) is
-        # returned, not the module named by module_name. However, when a
-        # non-empty fromlist argument is given, the module named by name is
-        # returned. Therefore, we set the source correctly here.
-        if not fromlist:
-            top_level_module_name = module_name.partition(".")[0]
-            source = self.import_source(top_level_module_name)
+        # Are we replaying? if so, load recorded module
+        recorded_name = ExecutionRecorder.LOCAL_MOD_PREFIX + module_name
+        if recorded_name in self.f_globals:
```

Comment:
What if we have multiple top level torchdynamo.optimize calls? Is it possible to get into a state where one leaves behind an artifact in globals, and then incorrectly causes the others to behave as if they are replaying?

Reply:
I don't think that's possible: when recording a DummyModule tree is built which doesn't modify f_globals.  Then when replaying this DummyModule tree is passed in to populate f_globals when calling torchdynamo.replay (this isn't the actual global environment, it's just a map). Now when import a module here, we check f_globals for it, and when we import a local module we also store it in f_globals with this prefix. The true global environment is never modified.